docs/shared/node-tutorial/4-task-pipelines.md

```diff
@@ -0,0 +1,235 @@
+
+{% card title="More On Customizing Inputs" description="See the Customizing Inputs Guide for more details on how to set inputs for your tasks." url="/concepts/task-pipeline-configuration" /%}
+
+### Outputs
+
+Outputs of the cache include the terminal output created by the task, as well as any files created by the task - for example: the artifact created by running a `build` task.
+
+Here are the outputs defined for the `auth` project:
+
+```json {% fileName="auth/project.json" %}
+{
+  "name": "shared-ui",
+  "$schema": "../../node_modules/nx/schemas/project-schema.json",
+  "projectType": "library",
+  "sourceRoot": "shared/ui/src",
+  "prefix": "store",
+  "targets": {
+    "build": {
+      "executor": "@nrwl/angular:ng-packagr-lite",
+      "outputs": ["{workspaceRoot}/dist/{projectRoot}"],
+      "options": {
+        "project": "shared/ui/ng-package.json"
+      },
+      "configurations": {
+        "production": {
+          "tsConfig": "shared/ui/tsconfig.lib.prod.json"
+        },
+        "development": {
+          "tsConfig": "shared/ui/tsconfig.lib.json"
+        }
+      },
+      "defaultConfiguration": "production"
+    },
+    "test": {
+      "executor": "@nrwl/jest:jest",
+      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
+      "options": {
+        "jestConfig": "shared/ui/jest.config.ts",
+        "passWithNoTests": true
+      }
+    },
+    "lint": {
+      "executor": "@nrwl/linter:eslint",
+      "options": {
+        "lintFilePatterns": ["shared/ui/**/*.ts", "shared/ui/**/*.html"]
+      }
+    }
+  },
+  "tags": []
+}
+```
+
+Outputs are stored in the cache so that terminal output can be replayed, and any created files can be pulled from your cache, and placed where they were created the original time the task was run.
+
+## Testing Affected Projects
+
+Another way that Nx saves you from unnecessary work is the `affected` command. `affected` is a mechanism that relies on your git metadata to determine the projects in your workspace that were affected by a given commit.
+
+Run the command:
+
+```shell
+git add . && git commit -m "commiting to test affected"
+```
+
+Then make a change to an endpoint of your `producst-api` project:
+
+```ts {% fileName="src/main.ts" %}
+import express from 'express';
+import { doAuth } from '@products-api/auth';
+
+const port = process.env.PORT ? Number(process.env.PORT) : 3000;
+
+const app = express();
+
+app.get('/', (req, res) => {
+  res.send({ message: 'Hello modified API' });
+});
+
+app.post('/auth', (req, res) => {
+  return doAuth();
+});
+
+app.listen(port, () => {
+  console.log(`[ ready ] http://localhost:${port}`);
+});
+```
+
+You can visualize how our workspace is affected by this change using the command:
+
+```shell
+npx nx affected:graph
+```
+
+{% graph height="450px" jsonFile="shared/node-tutorial/affected-project-graph.json" %}
+{% /graph %}
```

Comment:
The graph here does not seem to load, it shows a spinner indefinitely.
<img width="1253" alt="spinner" src="https://user-images.githubusercontent.com/3447705/219064373-2abac800-31f8-43a9-8cf6-bb8a46d4c2f4.png">